[BUG] Headless crash loop: Sharp Image Processor killed with exitCode 15 after health check

When running PostyBirb v4.0.41 in headless mode inside a Docker container, the application enters an infinite crash loop. The internal Sharp Image Processor utility child process passes its initial health check but is immediately terminated by the system with exitCode: 15 (SIGTERM), which forces the main container process to exit and restart.

Bumping shm_size or passing standard Electron/Chromium flags (like --disable-gpu or --no-sandbox) via environment variables does not prevent the child process from being killed./

postybirb-1  | [PostyBirb] Running in headless mode (no UI)
postybirb-1  | 2026.06.29 14:31:51.979 [DEBUG] [SettingsService] Added missing setting: tagSearchProvider.id
postybirb-1  | 2026.06.29 14:31:51.998 [INFO] Website registry marked as initialized
postybirb-1  | 2026.06.29 14:31:52.074 [INFO] Spawned sharp worker process { pid: 87, operation: 'healthcheck', isolatesCrashes: true }
postybirb-1  | 2026.06.29 14:31:52.448 [INFO] Sharp health check passed {
postybirb-1  |    platform: 'linux',
postybirb-1  |    arch: 'x64',
postybirb-1  |    nodeVersion: 'v24.15.0',
postybirb-1  |    sharpVersions: { ... },
postybirb-1  |    glibcVersion: '2.36',
postybirb-1  |    malloc_arena_max: '2'
postybirb-1  | }
postybirb-1  | 2026.06.29 14:31:52.460 [ERROR] [Diagnostics] Child process gone — type: Utility, reason: killed, exitCode: 15, service: node.mojom.NodeService, name: PostyBirb Sharp Image Processor
postybirb-1 exited with code 0 (restarting)


Environment & Context

    PostyBirb Version: 4.0.41

    Docker Image: ghcr.io/mvdicarlo/postybirb:latest

    Host OS: Linux x86_64

    Hardware: Intel Core i7-7700 / 32GB RAM

    Docker Compose Configuration:

YAML

services:
  postybirb:
    image: ghcr.io/mvdicarlo/postybirb:latest
    ports:
      - 8888:8888
    environment:
      - POSTYBIRB_PORT=8888
    volumes:
      - ./storage:/root/PostyBirb
      - ./appData:/root/.config/postybirb
    restart: unless-stopped

I can confirm that I see similar failures on Bazzite 44 with libvips errors. On startup the healthcheck fails with:
```
2026.06.29 17:31:31.137 [INFO] Spawned sharp worker process { pid: 5180, operation: 'healthcheck', isolatesCrashes: true }
2026.06.29 17:31:35.185 [INFO] [Updates] Checking for update
2026.06.29 17:31:36.741 [INFO] [Updates] Update for version 4.0.41 is not available (latest version: 4.0.41, downgrade is disallowed).
2026.06.29 17:31:38.951 [ERROR] [Diagnostics] Unhandled promise rejection in main process. Error: ERR_FAILED (-2) loading 'file:///tmp/.mount_PostyBv7ULj2/resources/app.asar/dist/apps/postybirb/app/loader/loader.html'
    at rejectAndCleanup (node:electron/js2c/browser_init:2:89743)
    at finishListener (node:electron/js2c/browser_init:2:89905)
    at WebContents.stopLoadingListener (node:electron/js2c/browser_init:2:90292)
    at WebContents.emit (node:events:521:24) {
  errno: -2,
  code: 'ERR_FAILED',
  url: 'file:///tmp/.mount_PostyBv7ULj2/resources/app.asar/dist/apps/postybirb/app/loader/loader.html'
}
2026.06.29 17:32:03.207 [ERROR] Sharp worker process exited before returning a result (likely a native libvips crash or out-of-memory) { operation: 'healthcheck', exitCode: 139, pid: undefined }
2026.06.29 17:32:03.208 [ERROR] Sharp health check FAILED — image processing will not work. This usually means native sharp bindings failed to load. On Linux, ensure glibc >= 2.17 is installed. On Snap/Flatpak, the sandbox may prevent loading native modules. Error: Sharp worker process exited unexpectedly (code 139) during healthcheck. This usually indicates a native libvips crash or out-of-memory while processing an image.
    at /tmp/.mount_PostyBv7ULj2/resources/app.asar/dist/apps/postybirb/main.js:3342:35
    at finish (/tmp/.mount_PostyBv7ULj2/resources/app.asar/dist/apps/postybirb/main.js:3288:17)
    at ForkUtilityProcess.<anonymous> (/tmp/.mount_PostyBv7ULj2/resources/app.asar/dist/apps/postybirb/main.js:3333:17)
    at ForkUtilityProcess.emit (node:events:509:28)
    at ForkUtilityProcess.a.emit (node:electron/js2c/browser_init:2:84209)
2026.06.29 17:32:03.209 [ERROR] [Diagnostics] Child process gone — type: Utility, reason: crashed, exitCode: 139, service: node.mojom.NodeService, name: PostyBirb Sharp Image Processor
```
On attempting to create a new submission:
```
2026.06.29 20:47:30.822 [INFO] [Mutation] Populating as Image
2026.06.29 20:47:30.826 [INFO] Spawned sharp worker process { pid: 12976, operation: 'metadata', isolatesCrashes: true }
2026.06.29 20:47:52.877 [ERROR] Sharp worker process exited before returning a result (likely a native libvips crash or out-of-memory) { operation: 'metadata', exitCode: 139, pid: undefined }
2026.06.29 20:47:52.877 [WARN] Rolling back transaction: cleaning up 1 entities
2026.06.29 20:47:52.890 [DEBUG] Cleaned up SubmissionFileSchema entity: 9e0f5d73-3244-486e-b113-a998c2a34e13
2026.06.29 20:47:52.890 [ERROR] undefined
2026.06.29 20:47:52.890 [ERROR] Error: Sharp worker process exited unexpectedly (code 139) during metadata. This usually indicates a native libvips crash or out-of-memory while processing an image.
    at /tmp/.mount_PostyBv7ULj2/resources/app.asar/dist/apps/postybirb/main.js:3342:35
    at finish (/tmp/.mount_PostyBv7ULj2/resources/app.asar/dist/apps/postybirb/main.js:3288:17)
    at ForkUtilityProcess.<anonymous> (/tmp/.mount_PostyBv7ULj2/resources/app.asar/dist/apps/postybirb/main.js:3333:17)
    at ForkUtilityProcess.emit (node:events:509:28)
    at ForkUtilityProcess.a.emit (node:electron/js2c/browser_init:2:84209) {
  level: 'error',
  Symbol(level): 'error',
  Symbol(message): '{"level":"error"}'
} Error creating submission
2026.06.29 20:47:52.904 [ERROR] [Diagnostics] Child process gone — type: Utility, reason: crashed, exitCode: 139, service: node.mojom.NodeService, name: PostyBirb Sharp Image Processor
```

--- a/apps/postybirb/src/app/app.ts
+++ b/apps/postybirb/src/app/app.ts
@@ -1,4 +1,3 @@
-import { INestApplication } from '@nestjs/common';
 import { app } from 'electron';
 import { installAppSecurity } from './main-process/security';
 import { TrayManager } from './main-process/tray';
@@ -13,8 +12,6 @@ export default class PostyBirbApp {
   private readonly windowManager = new WindowManager();
 
   private tray: TrayManager | null = null;
-
-  constructor(private readonly nestApp: INestApplication) {}
 
   /**
    * Install security policies, create the main window and tray, and register
@@ -35,8 +32,8 @@ export default class PostyBirbApp {
   }
 
   private registerLifecycleEvents(): void {
-    // Keep PostyBirb running in the system tray after the last window closes.
-    app.on('window-all-closed', () => {});
+    // The `window-all-closed` override is registered globally in main.ts so it
+    // applies in both GUI and headless modes (see the crash-loop note there).
 
     // macOS: re-show the window when the dock icon is clicked.
     app.on('activate', () => {
@@ -49,17 +46,5 @@ export default class PostyBirbApp {
         this.windowManager.show();
       }
     });
-
-    app.on('quit', () => {
-      this.shutdown();
-    });
-  }
-
-  private async shutdown(): Promise<void> {
-    await this.nestApp.close().catch((error) => {
-      // Ignore errors during shutdown
-      console.error('Error during shutdown:', error);
-    });
-    process.exit();
   }
 }

--- a/apps/postybirb/src/app/main-process/startup.ts
+++ b/apps/postybirb/src/app/main-process/startup.ts
@@ -2,9 +2,9 @@ import { INestApplication } from '@nestjs/common';
 import { PostyBirbDirectories } from '@postybirb/fs';
 import { flushAppInsights, Logger, trackException } from '@postybirb/logger';
 import {
-    PostyBirbEnvConfig,
-    RemoteConfigManager,
-    toError,
+  PostyBirbEnvConfig,
+  RemoteConfigManager,
+  toError,
 } from '@postybirb/utils/common';
 import { app } from 'electron';
 import { environment } from '../../environments/environment';
@@ -125,5 +125,57 @@ export function quitOnStartupFailure(error: unknown): void {
 
   flushAppInsights().finally(() => {
     app.quit();
+  });
+}
+
+/**
+ * Wire a single graceful-shutdown path for every launch mode. The embedded
+ * NestJS server must close cleanly — flushing the database and terminating its
+ * worker processes — however the quit is triggered:
+ *  - GUI: the tray "Quit" action or Cmd-Q drive `app.quit()`.
+ *  - Terminal (GUI or headless): Ctrl-C sends SIGINT and `docker stop` sends
+ *    SIGTERM. Electron does not turn these signals into a quit on its own, so
+ *    without this the process would be killed abruptly with no cleanup.
+ *
+ * Registered once, after the server has bootstrapped, so `nestApp` is always
+ * available when a quit is requested.
+ */
+export function registerGracefulShutdown(nestApp: INestApplication): void {
+  let isShuttingDown = false;
+
+  const closeAndExit = (): void => {
+    if (isShuttingDown) {
+      return;
+    }
+    isShuttingDown = true;
+
+    // Guard against a hung close leaving the container running indefinitely.
+    const forceExitTimer = setTimeout(() => app.exit(0), 5_000);
+    forceExitTimer.unref();
+
+    nestApp
+      .close()
+      .catch((error) => {
+        logger
+          .withError(toError(error))
+          .error('Error while closing the embedded server during shutdown.');
+      })
+      .finally(() => {
+        clearTimeout(forceExitTimer);
+        app.exit(0);
+      });
+  };
+
+  // Translate container/terminal termination signals into an Electron quit.
+  process.once('SIGTERM', () => app.quit());
+  process.once('SIGINT', () => app.quit());
+
+  // Defer the actual quit until the embedded server has closed cleanly.
+  app.on('before-quit', (event) => {
+    if (isShuttingDown) {
+      return;
+    }
+    event.preventDefault();
+    closeAndExit();
   });
 }

--- a/apps/postybirb/src/main.ts
+++ b/apps/postybirb/src/main.ts
@@ -22,11 +22,13 @@ import {
   registerProcessErrorHandlers,
 } from './app/main-process/diagnostics';
 import { startupLoader } from './app/main-process/loader';
+import { installAppSecurity } from './app/main-process/security';
 import {
   bootstrapWithTimeout,
   injectProcessEnvironment,
   logStartupBanner,
   quitOnStartupFailure,
+  registerGracefulShutdown,
 } from './app/main-process/startup';
 import { environment } from './environments/environment';
 
@@ -81,7 +83,17 @@ async function start(): Promise<void> {
   try {
     const nestApp = await bootstrapWithTimeout(bootstrapClientServer);
 
+    // A single graceful-shutdown path for every launch mode (GUI, terminal,
+    // headless/Docker). Registered here — after bootstrap — so the embedded
+    // server is always available to close cleanly when a quit is requested.
+    registerGracefulShutdown(nestApp);
+
     if (PostyBirbEnvConfig.headless) {
+      // Background login/scraper flows still open hidden BrowserWindows in
+      // headless mode, so apply the same security policies the GUI path applies
+      // via PostyBirbApp.
+      installAppSecurity();
+
       // eslint-disable-next-line no-console
       console.log('[PostyBirb] Running in headless mode (no UI)');
       return;
@@ -91,7 +103,7 @@ async function start(): Promise<void> {
     // can synchronously read app metadata as it loads.
     bootstrapElectronEvents();
 
-    const postyBirb = new PostyBirbApp(nestApp);
+    const postyBirb = new PostyBirbApp();
     postyBirb.start();
   } catch (error) {
     quitOnStartupFailure(error);
@@ -112,6 +124,17 @@ app
   .catch((error) => {
     quitOnStartupFailure(error);
   });
+
+// Keep the application alive when the last window closes. Electron's default
+// behavior quits the app once the open-window count reaches zero. In GUI mode
+// this keeps PostyBirb running in the system tray; in headless mode it is
+// essential — background login/scraper flows briefly open and close hidden
+// BrowserWindows, and the default quit-on-zero-windows behavior would tear down
+// the process (killing the NestJS server and its child processes), producing an
+// infinite crash/restart loop under Docker.
+app.on('window-all-closed', () => {
+  // Intentionally empty: overrides Electron's default quit-on-all-closed.
+});
 
 app.on('before-quit', () => {
   startupLoader.hide('before-quit');

